### DISCUSSION TITLE
비관적 lock, 낙관적 lock이 무엇일까요?

### DISCUSSION THREAD

**Discussion by Hanjaemo** [Hanjaemo/DB-Study#21] [Q&A]
비관적 lock, 낙관적 lock이 각각 무엇을 의미하는지 알려주세요.

**Comment by gomudayya** [comment]
**비관적 락**은 저희가 흔히 이야기하는 Read-Lock(shared), Write-Lock(exclusive)이야기 할 때의 그 Lock입니다.

**실제로 락을 점유하여 데이터의 동시성을 관리** 합니다.

실제 SQL로 예시를 들면 (MySQL) 아래와 같습니다.

```SQL
SELECT balance from  .... for update   // Write-Lock
SELECT balance from  .... for shared   // Read-Lock
```

SELECT로 조회를 하면서 해당 데이터에 락을 걸어버립니다. 보통 JPA와 같은 ORM을 사용하게 되면 SELECT로 먼저 조회를 하고, 데이터를 어플리케이션 단으로 가져온 이후에 변경하기 때문에 위와 같은 방법으로 사용하게 됩니다.

(참고로 UPDATE문만 사용해서 데이터를 변경한다면 UPDATE 쿼리 그 자체로 락이 걸려서 이런식으로도 동시성 문제를 해결하기도 하는 것 같아요. [관련링크](https://www.inflearn.com/questions/682421/%EB%B9%84%EA%B4%80%EC%A0%81-%EB%9D%BD%EA%B3%BC-update-product-set-stock-stock-1-where-id-1%EC%9D%98-%EC%B0%A8%EC%9D%B4))

반면에 **낙관적 락**은 **이름만 락이 들어가고, 실제 구현된 방법을 보면 락을 사용하지 않습니다.**

낙관적 락이 작동하는 방식은 아래와 같습니다.

1. 데이터를 업데이트하기 전에 해당 데이터의 버전을 기록합니다. (실제 테이블에 version column을 추가해서 사용함.)

2. 데이터를 업데이트하기 전에 다시 한 번 데이터를 확인합니다. 이때 다른 트랜잭션에 의해 수정되었다면 롤백됩니다.

   - 다른 트랜잭션에 의해 수정되었는지의 여부는 version값으로 판단합니다.

   - 예를들어 처음에 조회했을 때 version이 1이였는데, 업데이트 전에 다시 확인해보니 version이 2라면 다른 tx에 의해 수정된 것입니다.

3. 다른 트랜잭션에 의해 수정이 되지 않았다면 데이터를 업데이트하고, version도 업데이트 합니다.

어떤 상황에서 사용해야 하는지 판단해보면

**비관적락의 경우에는 동시성강도가 보다 높은상황**. 즉 동시성 이슈가 빈번하게 발생하는 상황에서 사용하는 것이 좋습니다.

왜냐하면 비관적 락은 무조건 락을 점유하기 때문에 기본적으로 실행 시간이 조금 더 느립니다.

반면 낙관적락은 좀 더 빠른편이죠. (그리고 얘는 이름 그대로 **동시성 이슈가 잘 발생하지 않을 것이라는 낙관적인 스탠스**입니다)

하지만 낙관적락의 경우 수정 실패시 tx가 롤백이 됩니다. 경우에 따라 실패했을 경우 retry를 해주어야 하는데, 이 때 실패한 트랜잭션이 엄청 많아서 이것들이 전부 retry해야되는 상황이라면 비관적락보다 비용이 많이 들게 됩니다.
